src/qunit.cpp

```diff
@@ -284,6 +286,91 @@ template <typename F, typename... B> void QUnit::EntangleAndCallMemberRot(F fn,
     ((*qbits).*fn)(radians, bits...);
 }
 
+template <typename CF, typename F>
+void QUnit::ControlCallMember(CF cfn, F fn, bitLenInt control, bitLenInt target, bool anti)
+{
+    real1 prob = Prob(control);
+    if (anti) {
+        prob = ONE_R1 - prob;
+    }
+    if (prob <= REAL_CLAMP) {
+        if (shards[control].unit->IsPhaseSeparable()) {
+            ForceM(control, anti);
+        }
+        return;
+    } else if (REAL_CLAMP >= (ONE_R1 - prob)) {
+        if (shards[control].unit->IsPhaseSeparable()) {
+            ForceM(control, !anti);
+        }
+        ((*(shards[target].unit)).*fn)(shards[target].mapped);
+        return;
+    }
+
+    bitLenInt tCopy = target;
+    auto qbits = Entangle({ &control, &target });
+    ((*qbits).*cfn)(control, target);
+    TrySeparate({ tCopy });
+}
+
+template <typename CF, typename F>
+void QUnit::ControlRotCallMember(CF cfn, F fn, real1 radians, bitLenInt control, bitLenInt target)
+{
+    real1 prob = Prob(control);
+    if (prob <= REAL_CLAMP) {
+        if (shards[control].unit->IsPhaseSeparable()) {
+            ForceM(control, false);
+        }
+        return;
+    } else if (REAL_CLAMP >= (ONE_R1 - prob)) {
+        if (shards[control].unit->IsPhaseSeparable()) {
+            ForceM(control, true);
+        }
+        ((*(shards[target].unit)).*fn)(radians, shards[target].mapped);
+        return;
+    }
+
+    bitLenInt tCopy = target;
+    auto qbits = Entangle({ &control, &target });
+    ((*qbits).*cfn)(radians, control, target);
+    TrySeparate({ tCopy });
+}
+
+void QUnit::TrySeparate(std::vector<bitLenInt> bits)
+{
+    for (bitLenInt i = 0; i < (bits.size()); i++) {
+        if (shards[bits[i]].unit->GetQubitCount() > 1) {
+            real1 oneChance = Prob(bits[i]);
+            if (oneChance <= REAL_CLAMP) {
+                if (shards[bits[i]].unit->IsPhaseSeparable()) {
+                    ForceM(bits[i], false);
+                }
+            } else if (oneChance >= (ONE_R1 - REAL_CLAMP)) {
+                if (shards[bits[i]].unit->IsPhaseSeparable()) {
+                    ForceM(bits[i], true);
+                }
+            }
+        }
+    }
+}
+
+void QUnit::TrySeparate(bitLenInt start, bitLenInt length)
```

Comment:
Maybe implement this one in terms of the vector<> one?

Reply:
Better yet, we remove it, for the moment. INC/DEC variants don't use it all, for now. As for why, I talked it out with myself in the Discord channel, if you want to try to read that, but this hurts performance and isn't the long-term solution. Let's just remove it, for now.